libs/algo/odc/algo/_tiff.py

```diff
@@ -0,0 +1,518 @@
+                break
+
+            ii = ii.shrink2()
+            ext = ext + ".ovr"
+            if bsz > 64:
+                bsz = bsz // 2
+
+        self._layers = layers
+        self._mem = MemoryFile() if dst == ":mem:" else None
+        self._dst = dst
+        self._rio_opts = opts
+        self._ovr_blocksize = ovr_blocksize
+        self._resampling = overview_resampling
+        self._info = info
+
+    def _shrink2(self, xx, roi):
+        axis = self._info.axis
+        out_roi = roi_shrink2(roi, axis=axis)
+        out = _shrink2(
+            xx, resampling=self._resampling, nodata=self._info.nodata, axis=axis
+        )
+
+        return out_roi, out
+
+    def __setitem__(self, key: NumpyIndex, item: np.ndarray):
+        dst, *ovrs = self._layers
+        dst[key] = item
+        for dst in ovrs:
+            key, item = self._shrink2(item, key)
+            dst[key] = item
+
+    def close(self, idx=-1):
+        if idx < 0:
+            for dst in self._layers:
+                dst.close()
+        elif idx < len(self._layers):
+            self._layers[idx].close()
+
+    def _copy_cog(self, extract=False, strict=False) -> Optional[bytes]:
+        with rasterio.Env(
+            GDAL_TIFF_OVR_BLOCKSIZE=self._ovr_blocksize,
+            GDAL_DISABLE_READDIR_ON_OPEN=False,
+            NUM_THREADS="ALL_CPUS",
+            GDAL_NUM_THREADS="ALL_CPUS",
+        ):
+            src = self._layers[0].name
+            if self._mem is not None:
+                rio_copy(
+                    src,
+                    self._mem.name,
+                    copy_src_overviews=True,
+                    strict=strict,
+                    **self._rio_opts,
+                )
+                if extract:
+                    # NOTE: this creates a copy of compressed bytes
+                    return bytes(self._mem.getbuffer())
+            else:
+                rio_copy(
+                    src,
+                    self._dst,
+                    copy_src_overviews=True,
+                    strict=strict,
+                    **self._rio_opts,
+                )
+            return None
+
+    def finalise(self, extract=False, strict=False) -> Optional[bytes]:
+        self.close()  # Write out any remainders if needed
+        return self._copy_cog(strict=strict, extract=extract)
+
+    def mem(self):
+        return self._mem
+
+    def dump_to_s3(self, url, **kw):
+        import boto3
+        from boto3.s3.transfer import TransferConfig
+        from odc.aws import s3_url_parse
+
+        assert self._mem is not None
+
+        GB = 1 << 30
+        transfer_config = TransferConfig(multipart_threshold=5 * GB)
+        bucket, key = s3_url_parse(url)
+        s3 = boto3.client("s3")
+
+        return s3.upload_fileobj(
+            self._mem, bucket, key, ExtraArgs=kw, Config=transfer_config
+        )
+
+    @staticmethod
+    def dask_finalise(sink: Delayed, extract=False, strict=False, *deps) -> Delayed:
```

Comment:
why is this a static method? is it something to do with needed to wrap the `COGSink` object with `Delayed`?

Reply:
It's not operating on `self`, `self` wasn't even created yet at this point.

Reply:
oh yep I see that now, still wrapping my head around dask! but why do you use this function as opposed to doing something like `dask.delayed(lambda sink, *deps: sink.finalise())`?